### ISSUE TITLE
Improve our Pimpl approach?

### ISSUE THREAD

**Issue by agarny** [cellml/libcellml#279]
Our current implementation of the Pimpl idiom relies on each of our classes having a raw pointer to its private implementation. This means that our `Component` class has 5 private implementations, one for itself, another for its parent classes, i.e. `ComponentEntity`, `NamedEntity`, `Entity`, and a final one for `ImportedEntity` since `Component` inherits both from `ComponentEntity` and `ImportedEntitiy`.

However, for such a class, we could get away with "only" 2 private impelementations, one for `ComponentEntity`-`NamedEntity`-`Entity` and another for `ImportedEntity`, this by using inheritance.

~~Also, we could use a smart pointer rather than a raw pointer for our private implementations.~~

In other words, we might want to have a look at [what the Qt guys are doing](https://wiki.qt.io/D-Pointer) and probably mimick their implementation of the Pimpl idiom...?

P.S.: we might also want to clean some of our interfaces. E.g. the `private` section of the `Component` class contains the `doAddComponent()` method, which I believe has nothing to do there and could probably be "fixed" by using inheritance for our private implementations.

**Comment by MichaelClerx** [comment]
(I'm not too involved in the C++ side, but googling shows Pimpl is mainly good for fast compilation. Is that the reason we're using it here, or is libcellml too small to worry about that?)

**Comment by nickerso** [comment]
Pretty sure this approach was taken to keep the API clean as the main reason. @hsorby will likely correct me :)

**Comment by agarny** [comment]
I am no Pimpl expert, but I understand that it's also good for binary compatibility, among other things, which is quite important when it comes to APIs such as libCellML. There are clearly advantages and disadvantages to implementing the Pimpl idiom, but I think that's besides the point here. For some reasons (I don't think I was involved in that decision), it was agreed that libCellML would implement the Pimpl idiom, so I am merely pointing out that our implementation could, I think, be improved.

**Comment by hsorby** [comment]
The Pimpl idiom is being used mainly for the point of being able to control the API.  But there are other benefits like faster compilation.

Part of the keep the API clean is for binary compatibility reasons with the Pimpl idiom it is easier to add methods and not break binary compatibility.

The doAddComponent might be a hangover from when the model object had a different inheritance hierarchy.  This is a virtual function that should not to be in the public API.

I think the Qt guys are using raw pointers as well???

**Comment by agarny** [comment]
> I think the Qt guys are using raw pointers as well???

My bad, the Qt guys do indeed use raw pointers, and rightly so since there is no point in using smart pointers in that context.

**Comment by agarny** [comment]
Ok, after trying to improve our Pimpl approach, I have to admit defeat. Several issues make it difficult to take advantage of inheritance for our private implementations. For example, we have copy/move constructors, assignment operator and the swap method which makes things a bit more difficult. Also, and that is the killer, we use multiple-inheritance. So, to have inheritance for our private implementations means that a class like `Component` would end up with two private implementations, making it rather difficult to manage.

Bottom line: closing this issue...
